lib/xgu/xgu.h

```diff
@@ -0,0 +1,691 @@
+#ifndef __XGU_H
+#define __XGU_H
+
+#include <assert.h>
+#include <pbkit/pbkit.h>
+// Hack until we fix pbkit/outer.h (or stop including it)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmacro-redefined"
+#include "nv2a_regs.h"
+#pragma GCC diagnostic pop
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <xboxrt/strings.h>
+#include <string.h>
+
+typedef enum {
+    //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_UB_D3D     0
+    //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_S1         1
+    XGU_FLOAT = NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_F,
+    //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_UB_OGL     4
+    //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_S32K       5
+    //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_CMP        6
+} XguVertexArrayType;
+
+typedef enum {
+    XGU_END = NV097_SET_BEGIN_END_OP_END, // FIXME: Disallow this one?
+    XGU_POINTS = NV097_SET_BEGIN_END_OP_POINTS,
+    XGU_LINES = NV097_SET_BEGIN_END_OP_LINES,
+    XGU_LINE_LOOP = NV097_SET_BEGIN_END_OP_LINE_LOOP,
+    XGU_LINE_STRIP = NV097_SET_BEGIN_END_OP_LINE_STRIP,
+    XGU_TRIANGLES = NV097_SET_BEGIN_END_OP_TRIANGLES,
+    XGU_TRIANGLE_STRIP = NV097_SET_BEGIN_END_OP_TRIANGLE_STRIP,
+    XGU_TRIANGLE_FAN = NV097_SET_BEGIN_END_OP_TRIANGLE_FAN,
+    XGU_QUADS = NV097_SET_BEGIN_END_OP_QUADS,
+    XGU_QUAD_STRIP = NV097_SET_BEGIN_END_OP_QUAD_STRIP,
+    XGU_POLYGON = NV097_SET_BEGIN_END_OP_POLYGON
+} XguPrimitiveType;
+
+typedef enum {
+    XGU_VERTEX_ARRAY = NV2A_VERTEX_ATTR_POSITION,
+    XGU_NORMAL_ARRAY = NV2A_VERTEX_ATTR_NORMAL,
+    XGU_COLOR_ARRAY = NV2A_VERTEX_ATTR_DIFFUSE,
+    XGU_SECONDARY_COLOR_ARRAY = NV2A_VERTEX_ATTR_SPECULAR,
+    XGU_FOG_ARRAY = NV2A_VERTEX_ATTR_FOG,
+    XGU_TEXCOORD0_ARRAY = NV2A_VERTEX_ATTR_TEXTURE0,
+    XGU_TEXCOORD1_ARRAY = NV2A_VERTEX_ATTR_TEXTURE1,
+    XGU_TEXCOORD2_ARRAY = NV2A_VERTEX_ATTR_TEXTURE2,
+    XGU_TEXCOORD3_ARRAY = NV2A_VERTEX_ATTR_TEXTURE3
+} XguVertexArray;
```

Comment:
nv2a_regs.h is affected by XQEMU issue 258 [not linked to avoid github reference in that XQEMU issue]; this error propagates into xgu.h constants `XGU_TEXCOORD0_ARRAY` etc.

Reply:
I think XQEMU is actually fine, but our shader compilers are broken. I'm marking this resolved.